🚧 gRPC Support with .NET Core 3.0

jurby opened the pull request:
Add gRPC Support with .NET Core 3.0

- [x] Add `AspNetCore.Grpc`
- [x] Redefine grpc graphql API
- [x] Implement `GraphqlService.Query`
- [x] Implement `GraphqlService.QueryBatch`
- [ ] Implement mutations over gRPC `GraphqlService.Mutation`
- [ ] Implement subscriptions over gRPC `GraphqlService.Subscription`
- [x] Extensions `AddGraphQLOverGrpc `and `UseGraphQLOverGrpc`.
- [x] Optimalization: avoid using `HotChocolate.Execution.IExecutionResult` -> `Json` -> `Google.Protobuf.WellKnownTypes.Struct` conversion; use `HotChocolate.Execution.IExecutionResult` -> `Google.Protobuf.WellKnownTypes.Struct` -> make extension `HotChocolate.Execution.IExecutionResult.ToStruct()`
- [x] Add AspNetCore.Grpc.Abstractions - contains server, client and contracts
- [ ] When using gRPC services via `endpoints.MapGrpcService<GraphqlGrpcService>()` than not working routing for GraphQL / GraphiQL / Playground / Voyager.
- [ ] Refactoring - code formatting and conventions 
- [ ] Update documentation

Addresses #602

jurby commented:
[TBD] When using gRPC services via endpoints.MapGrpcService <GraphqlGrpcService> () than not working routing for GraphQL / GraphiQL / Playground / Voyager.

jurby commented on examples/AspNetCore.StarWars/Startup.cs:
```diff
@@ -80,14 +91,31 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+            // Enable HTTP2 only on HTTPS
+            app.UseHttpsRedirection();
+            app.UseDefaultFiles();
+            app.UseStaticFiles();
```
No. I will remove them.

jurby commented on examples/AspNetCore.StarWars/Startup.cs:
```diff
@@ -80,14 +91,31 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+
+            app.UseEndpoints(endpoints =>
+            {
+                endpoints.MapGrpcService<GraphqlGrpcService>();
```
I make extension `AddGraphQLOnGrpc` and `UseGraphQLOnGrpc`.

jurby commented on examples/AspNetCore.StarWars/Services/TestGrpcToGraphqlHostedService.cs:
```diff
@@ -0,0 +1,124 @@
+    /// <summary>
+    /// Test the gRPC client with the gRPC GraphQL service
+    /// </summary>
+    public class TestGrpcToGraphqlHostedService
```
OK. I didn't know where to put it to see the use. I didn't want to create another project just as a test client.

michaelstaib replied:
I think I will add gRPC to our client soon, then you will be able to use that to test the service

jurby replied:
OK

jurby commented on examples/AspNetCore.StarWars/appsettings.json:
```diff
@@ -0,0 +1,11 @@
+{
```
OK

jurby commented on src/Server/AspNetCore.Grpc/Extensions/ExecutionResultExtensions.cs:
```diff
@@ -0,0 +1,60 @@
+            var response = new Response
+            {
+                // TODO: Must be reworked to avoid using HotChocolate.Execution.IExecutionResult -> Json -> Google.Protobuf.WellKnownTypes.Struct conversion
+                Data = JsonParser.Default.Parse<Struct>(result.ToJson()),
```
I'll try to prepare it.

jurby replied:
Finished

jurby commented:
> We should talk about how the external interface of gRPC should look like.

I agree. In my opinion, it should match what can be called using the HTTP API.

> Also we need to get the serialization and deserialization right.

I'll try to prepare it.

> Good start.

Thank you.

> I will look a bit more closely this week and then we will redefine the interface a bit...

OK

>...I think we should also introduce subscriptions over gRPC.

I agree.

jurby commented on examples/AspNetCore.StarWars/Startup.cs:
```diff
@@ -80,14 +91,31 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             }
 
+            // Enable HTTP2 only on HTTPS
+            app.UseHttpsRedirection();
```
gRPC clients need to use HTTPS to call the server - https://docs.microsoft.com/en-us/aspnet/core/tutorials/grpc/grpc-start?view=aspnetcore-3.0&tabs=visual-studio#run-the-service

jurby commented:
@michaelstaib I added and improved some parts. I modified the definition of the grpc API.

@michaelstaib I would like to ask you how do I get the Path property to send the Query to a client? The point is that the answer does not have to be received at the same time (eg @defer or @stream) or asynchronous streams with answers (eg QueryResponse), where it is necessary to return also the path (Path).
See. https://www.youtube.com/watch?v=ViXL0YQnioU&feature=youtu.be&t=626

michaelstaib commented:
@jurby we internally have the path at the moment but the new execution engine is not in yet, so we do not have it on the result yet. We will add the the path property to the IReadOnlyQueryResult. Path can be null if it is not a patch request.

It will still take some time for these things to come in. We first will start on the resolver compiler to expose the compiler metadata so that the execution plan  feature has all the needed data to make decisions on what resources are needed for each resolver. 

Thinking about this, I probably could add this property for you but it will for now always be null, this way I would not block you.

jurby commented:
> Thinking about this, I probably could add this property for you but it will for now always be null, this way I would not block you.

@michaelstaib Thanks a lot for the information. I assumed it wasn't ready yet. I went through the code and did not find it;)
It would be ideal to add the `Path` property to `IReadOnlyQueryResult`. But it is not absolutely necessary. I will leave it blank (null) for now. Thank you.

For code formatting and conventions, I will customize them according to what you wrote.

May I ask what is the plan for v11 - when is it expected?

michaelstaib commented:
@jurby originally I thought end of November but I think end of. January 2020 is more realistic. We have still so much to do for the new execution engine, the new client API, the filters and the ASP.NET middleware. So, your work is well appreciated.

michaelstaib commented:
Hi @jurby,

We are now starting work on the server for v11 and will soon start to integrate your work.
